src/node/diagnostics_channel.ts

```diff
@@ -0,0 +1,243 @@
+      } catch (err) {
+        (context as any).error = err;
+        error.publish(context);
+        throw err;
+      } finally {
+        end.publish(context);
+      }
+    }, thisArg);
+  }
+
+  tracePromise(fn : (...args: any[]) => any,
+            context: Object = {},
+            thisArg: any = globalThis,
+            ...args: any[]) {
+    const { start, end, asyncStart, asyncEnd, error } = this;
+
+    function reject(err: any) {
+      (context as any).error = err;
+      error.publish(context);
+      asyncStart.publish(context);
+      asyncEnd.publish(context);
+      return Promise.reject(err);
+    }
+
+    function resolve(result: any) {
+      (context as any).result = result;
+      asyncStart.publish(context);
+      asyncEnd.publish(context);
+      return result;
+    }
+
+    return start.runStores(context, () => {
+      try {
+        let promise = Reflect.apply(fn, thisArg, args);
+        // Convert thenables to native promises
+        if (!(promise instanceof Promise)) {
+          promise = Promise.resolve(promise);
+        }
+        return promise.then(resolve, reject);
+      } catch (err) {
+        (context as any).error = err;
+        error.publish(context);
+        throw err;
+      } finally {
+        end.publish(context);
+      }
+    }, thisArg);
+  }
+
+  traceCallback(fn : (...args: any[]) => any,
+                position = -1,
+                context = {},
+                thisArg: any = globalThis,
+                ...args: any[]) {
+    const { start, end, asyncStart, asyncEnd, error } = this;
+
+    function wrappedCallback(this: any, err: any, res: any) {
+      if (err) {
+        (context as any).error = err;
+        error.publish(context);
+      } else {
+        (context as any).result = res;
+      }
+
+      // Using runStores here enables manual context failure recovery
+      asyncStart.runStores(context, () => {
+        try {
+          if (callback) {
+            return Reflect.apply(callback, this, arguments);
+          }
+        } finally {
+          asyncEnd.publish(context);
+        }
+      }, thisArg);
+    }
+
+    const callback = args[position];
+    if (typeof callback !== 'function') {
+      throw new ERR_INVALID_ARG_TYPE('callback', ['function'], callback);
+    }
+    args.splice(position, 1, wrappedCallback);
+
+    return start.runStores(context, () => {
+      try {
+        return Reflect.apply(fn, thisArg, args);
+      } catch (err) {
+        (context as any).error = err;
+        error.publish(context);
+        throw err;
+      } finally {
+        end.publish(context);
+      }
+    }, thisArg);
+  }
+}
+
+function validateChannel(channel: any, name: string) {
+  if (!(channel instanceof Channel)) {
+    throw new ERR_INVALID_ARG_TYPE(name, 'Channel', channel);
+  }
+  return channel as ChannelType;
+}
+
+export function tracingChannel(nameOrChannels : string|TracingChannels) {
+  return Reflect.construct(function (this: TracingChannel) {
```

Comment:
largely just because it's not clear from the Node.js documentation if that's supposed to work. I know it does if we grab the `TracingChannel` constructor but is that intentional?

e.g., the following works in Node.js but is undocumented since the `tracingChannel()` method is documented as the mechanism for acquiring a `TracingChannel` instance

```
const { tracingChannel } = require('diagnostics_channel');
const TracingChannel = tracingChannel('foo').constructor;
new TracingChannel('foo');  // works, but should it?

```

Reply:
Discussed privately with James and I'm going to make a PR to Node.js to align with this pattern. Constructing Channel and TracingChannel directly was never _intended_, but just happened to be possible.